ApeWorX/ape
# feat: add `snapshot()` and `revert()` implementations to `ape-geth` local provider

### What I did

While working on my trace call work, I realized I needed to separate the Geths in order to subscribe to `TestProviderAPI`.
While making the `GethDev` class, I noticed there were a couple Test APIs that  could easily be implemented, namely `snapshot()` and `revert()`.

### How I did it

Use `debug_setHead` to return to an older block.
For the `SnapshotID`, use the block number of the latest block.
Now, you have the same feature implemented in the Geth Dev provider we have.
This is 1 step closer to making GethDev great, because it has amazing tracing features so I want to encourage its use.

The work related to the refactor is needed for my trace work.

### How to verify it

**NOTE: Cannot add very interesting tests because the logic is rather simple and we don't have a `geth` installed on CI/CD yet to test on.**

Here is the test if I were to add one:

```python
def test_snapshot_and_revert(geth_dev, account_0, account_1):
    block_number = geth_dev.snapshot()
    # Increase blocks
    account_0.transfer(account_1, "1 ETH")
    assert geth_dev.get_block("latest") == block_number + 1
    geth_dev.revert(block_number)
    assert geth_dev.get_block("latest") == block_number
```

You can use or something like it to manually verify the feature.

### Checklist
